test/puppetlabs/puppetdb/testutils/jetty.clj

```diff
@@ -109,16 +110,121 @@
+                                                                  :throw-exceptions false})]
+        (cond
+
+         (= 200 status)
+         true
+
+         (<= max-attempts attempts)
+         (throw+ response "JMX not up after %s attempts" attempts)
+
+         :else
+         (do
+           (Thread/sleep 100)
+           (recur (inc attempts))))))))
+
+(defn queue-metadata
+  "Return command queue metadata (from the `puppetdb-instance` launched PuppetDB) as a map:
+
+  EnqueueCount - the total number of messages sent to the queue since the last restart
+  DequeueCount - the total number of messages removed from the queue (ack'd by consumer) since last restart
+  InflightCount - the number of messages sent to a consumer session and have not received an ack
+  DispatchCount - the total number of messages sent to consumer sessions (Dequeue + Inflight)
+  ExpiredCount - the number of messages that were not delivered because they were expired
+
+  http://activemq.apache.org/how-do-i-find-the-size-of-a-queue.html"
+  []
+  ;; When starting up a `puppetdb-instance` there seems to be a brief
+  ;; period of time that the server is up, the broker has been
+  ;; started, but the JMX beans have not been initialized, so querying
+  ;; for queue metrics fails, this check ensures it's started
   (let [base-metrics-url (assoc *base-url* :prefix "/metrics" :version :v1)]
     (-> (str (utils/base-url->str base-metrics-url)
              "/mbeans/org.apache.activemq:BrokerName="
              (url-encode (:host base-metrics-url))
-             ",Type=Queue,Destination=puppetlabs.puppetdb.commands")
+             ",Type=Queue,Destination="
+             mq-endpoint)
         (client/get {:as :json})
-        (get-in [:body :QueueSize]))))
+        :body)))
+
+(defn current-queue-depth
+  "Return the queue depth currently running PuppetDB instance (see `puppetdb-instance` for launching PuppetDB)"
+  []
+  (:QueueSize (queue-metadata)))
+
+(defn discard-count
+  "Return the number of discarded messages from the command queue for the current running `puppetdb-instance`"
+  []
+  (let [base-metrics-url (assoc *base-url* :prefix "/metrics" :version :v1)]
+    (-> (str (utils/base-url->str base-metrics-url)
+             "/mbeans/puppetlabs.puppetdb.command:type=global,name=discarded")
+        (client/get {:as :json})
+        (get-in [:body :Count]))))
+
+(defn until-consumed
+  "Invokes `f` and blocks until the `num-messages` have been consumed
+  from the commands queue. `num-messages` defaults to 1 if not
+  provided. Returns the result of `f` if successful. Requires JMX to
+  be enabled in ActiveMQ (the default, but `without-jmx` will cause
+  this to fail).
+
+  Exceptions thrown in the following cases:
+
+  timeout - if the message isn't consumed in approximately 5 seconds
+  new message in the DLO - if any message is discarded"
+  ([f] (until-consumed 1 f))
+  ([num-messages f]
```

Comment:
Oh, but you can't have varargs and multiple arities, we can support it on the second arity though.